halo2_proofs/src/dev/failure.rs

```diff
@@ -545,79 +546,81 @@ fn render_lookup<F: FieldExt>(
     eprintln!(")");
 
     eprintln!();
-    eprintln!("  Lookup '{}' inputs:", name);
-    for (i, input) in lookup.input_expressions.iter().enumerate() {
-        // Fetch the cell values (since we don't store them in VerifyFailure::Lookup).
-        let cell_values = input.evaluate(
-            &|_| BTreeMap::default(),
-            &|_| panic!("virtual selectors are removed during optimization"),
-            &cell_value(&util::load_slice(
-                n,
-                row,
-                &cs.fixed_queries,
-                prover.fixed.as_slice(),
-            )),
-            &cell_value(&util::load_slice(
-                n,
-                row,
-                &cs.advice_queries,
-                &prover.advice,
-            )),
-            &cell_value(&util::load_instance(
-                n,
-                row,
-                &cs.instance_queries,
-                &prover.instance,
-            )),
-            &|_| BTreeMap::default(),
-            &|a| a,
-            &|mut a, mut b| {
-                a.append(&mut b);
-                a
-            },
-            &|mut a, mut b| {
-                a.append(&mut b);
-                a
-            },
-            &|a, _| a,
-        );
-
-        // Collect the necessary rendering information:
-        // - The columns involved in this constraint.
-        // - How many cells are in each column.
-        // - The grid of cell values, indexed by rotation.
-        let mut columns = BTreeMap::<metadata::Column, usize>::default();
-        let mut layout = BTreeMap::<i32, BTreeMap<metadata::Column, _>>::default();
-        for (i, (cell, _)) in cell_values.iter().enumerate() {
-            *columns.entry(cell.column).or_default() += 1;
-            layout
-                .entry(cell.rotation)
-                .or_default()
-                .entry(cell.column)
-                .or_insert(format!("x{}", i));
-        }
+    eprintln!("  Lookup inputs:");
+    for input_expressions in lookup.inputs_expressions.iter() {
```

Comment:
What is happening at the lines 549-625? Why are the "a.append(&mut b);" lines provided instead of arithmetic operations?

Reply:
Oh it just collects all the data to print out the information as given in the example here: https://github.com/taikoxyz/halo2/blob/a265b2ca5d9e586bec9306aea4115b4e49c6b26f/halo2_proofs/src/dev/failure.rs#L434-L444.

So it parses the input expression to collect all the cell values that are used. So it doesn't actually try to calculate the input expression (so no arithmetic), it just prints all the used cell values for debugging purposes.

Reply:
Thanks! Just noticed that the data type of the value being appended is not a field element, but key-value pair:
![image](https://github.com/taikoxyz/halo2/assets/77882392/5763d15f-e885-4a85-ad71-c059a3919e52)
Thus, this makes sense for me now.